curp/src/sync_manager.rs

```diff
@@ -144,137 +139,233 @@ impl<C: Command + 'static> SyncManager<C> {
-                    Some(SyncResponse::Synced(_)) => 1,
-                    Some(
-                        SyncResponse::WrongTerm(_)
-                        | SyncResponse::EntryNotEmpty(_)
-                        | SyncResponse::PrevNotReady(_),
-                    ) => 0,
-                    None => unreachable!("Should contain sync response"),
-                },
-                comp_fn,
-                max_fail,
-            )
-            .await
-            {
-                return;
+                    })
+                    .collect();
+                let _drop = tokio::spawn(async move {
+                    let _drop: Vec<_> = rpcs.collect().await;
+                });
             }
+        }
+    }
 
-            let connects = self.connects();
-            let _ignore_handle = tokio::spawn(async move {
-                Self::broadcast_and_wait(
-                    connects,
-                    iter::repeat_with(|| Arc::clone(&cmds_arc)),
-                    |(connect, cmds_cloned)| async move {
-                        connect
-                            .commit(CommitRequest::new(
-                                term,
-                                index.numeric_cast(),
-                                cmds_cloned.as_ref(),
-                            )?)
-                            .await
-                    },
-                    |r| match r.into_inner().commit_response {
-                        Some(CommitResponse::Committed(_)) => 1,
-                        Some(CommitResponse::WrongTerm(_) | CommitResponse::PrevNotReady(_)) => 0,
-                        None => unreachable!("Should contain sync response"),
-                    },
-                    || true,
-                    max_fail,
+    /// Send heartbeat to a server
+    #[allow(clippy::integer_arithmetic, clippy::indexing_slicing)] // log.len() >= 1 because we have a fake log[0], indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+    async fn send_append_entries(
+        connect: Arc<Connect>,
+        state: Arc<RwLock<State<C>>>,
+        commit_trigger: UnboundedSender<()>,
+        ae_trigger: UnboundedSender<()>,
+    ) {
+        // prepare append_entries request args
+        #[allow(clippy::shadow_unrelated)] // clippy false positive
+        let (term, leader_id, prev_log_index, prev_log_term, entries, leader_commit) = state
+            .map_read(|state| {
+                let next_index = state.next_index[&connect.addr];
+                (
+                    state.term,
+                    0, // TODO: add leader_id
+                    next_index - 1,
+                    state.log[next_index - 1].term(),
+                    state.log[next_index..].to_vec(),
+                    state.commit_index,
                 )
-                .await
             });
+        let last_sent_index = prev_log_index + entries.len();
```

Comment:
There may be several `send_append_entries` called concurrently, the smaller `last_sent_index` doesn't mean the request is sent earlier as there's no lock protection. Please double check the correctness.

Reply:
The variable actually means the index of the last sent log entry **in the current request**. I will add a comment above this line of code.

Reply:
It's used in the resp handling. If the response with "last_sent_index 1" returns firstly, then "last_sent_index 0" response arrives. Will the next index and match index be set correctly?

Reply:
I see. I will add a check before changing match_index and next_index